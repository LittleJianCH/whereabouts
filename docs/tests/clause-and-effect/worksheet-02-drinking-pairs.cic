Drink { person: "john", alcohol: "martini" }
Drink { person: "mary", alcohol: "gin" }
Drink { person: "susan", alcohol: "vodka" }
Drink { person: "john", alcohol: "gin" }
Drink { person: "fred", alcohol: "gin" }
Drink { person: "fred", alcohol: "vodka" }

Friends { left, right, alcohol }
------------------------------------ {
  Drink { person: left, alcohol }
  Drink { person: right, alcohol }
}

query (left) {
  Friends { left, right: "mary", alcohol: "gin" }
}

query (left, right) {
  Friends { left, right, alcohol: "gin" }
}

query (left, right, alcohol) {
  Friends { left, right, alcohol }
}

console.log(
  Logic.query(
    (v) => [pair.o({ p1: v`x`, p2: v`y`, alcohol: v`alcohol` })],
    ["x", "y"]
  )
)

{
  type Result = { x: string; y: string }
  const results: Array<Result> = Logic.find(
    (v) => [pair.o({ p1: v`x`, p2: v`y`, alcohol: v`alcohol` })],
    { x: ty.string(), y: ty.string() }
  )
  console.log(results)
}

{
  type Result = { x: string }
  const results: Array<Result> = pair.find(
    { p1: v`x`, p2: v`y`, alcohol: v`alcohol` },
    { x: ty.string() }
  )
  console.log(results)
}
