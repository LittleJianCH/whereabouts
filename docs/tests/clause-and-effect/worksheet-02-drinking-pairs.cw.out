{ "solutions": [{ "left": "mary" }, { "left": "john" }, { "left": "fred" }] }
{ "solutions": [{ "left": "mary", "right": "mary" }, { "left": "mary", "right": "john" }, { "left": "mary", "right": "fred" }, { "left": "john", "right": "mary" }, { "left": "john", "right": "john" }, { "left": "john", "right": "fred" }, { "left": "fred", "right": "mary" }, { "left": "fred", "right": "john" }, { "left": "fred", "right": "fred" }] }
{ "solutions": [{ "left": "john", "right": "john", "alcohol": "martini" }, { "left": "mary", "right": "mary", "alcohol": "gin" }, { "left": "mary", "right": "john", "alcohol": "gin" }, { "left": "mary", "right": "fred", "alcohol": "gin" }, { "left": "susan", "right": "susan", "alcohol": "vodka" }, { "left": "susan", "right": "fred", "alcohol": "vodka" }, { "left": "john", "right": "mary", "alcohol": "gin" }, { "left": "john", "right": "john", "alcohol": "gin" }, { "left": "john", "right": "fred", "alcohol": "gin" }, { "left": "fred", "right": "mary", "alcohol": "gin" }, { "left": "fred", "right": "john", "alcohol": "gin" }, { "left": "fred", "right": "fred", "alcohol": "gin" }, { "left": "fred", "right": "susan", "alcohol": "vodka" }, { "left": "fred", "right": "fred", "alcohol": "vodka" }] }
